### PR TITLE
Peterborough Bulky Goods cancellations

### PR DESCRIPTION
**Pull request by nephila-nacrea**
Closes https://github.com/mysociety/societyworks/issues/2915.

[skip changelog]

### CODE REVIEW

**Review by dracos — CHANGES_REQUESTED** [review]
The workflow looks good, though really needs a test. Some other comments about things that can be removed, need a bit more work, or hopefully simplified. Please do ask if you have any questions about anything :)

File: t/cobrand/peterborough/bulky-goods/unit.t [hunk 1]
```diff
@@ -375,4 +376,318 @@ sub _jobs_arrayref {
     return [ map { { Job => { UPRN => $uprn++ } } } 1 .. $job_count ];
 }
 
+### end 'find_available_bulky_slots'
+
+subtest '_check_within_bulky_cancel_window' => sub {
```
[the remark this answers is not in the document]
Comment (nephila-nacrea) [reply t/cobrand/peterborough/bulky-goods/unit.t:381]: I am aware I can spend too much time on little details. I guess from where I was, I don't feel confident about DateTime maths / comparisons; I am always worried Weird Stuff Will Happen if I don't test it.

File: perllib/Integrations/Bartec.pm [hunk 2]
```diff
@@ -322,6 +322,22 @@ sub ServiceRequests_Get {
     return force_arrayref($requests, 'ServiceRequest');
 }
 
+# XXX Move to open311-adapter?
```
[the remark this answers is not in the document]
Comment (nephila-nacrea) [reply perllib/Integrations/Bartec.pm:331]: Oops, no it isn't.

File: templates/web/base/waste/bulky/booking_cancellation.html [hunk 3]
```diff
@@ -0,0 +1,24 @@
+<!-- Step 10.3D Wireframe Cancellation -->
```
[the remark this answers is not in the document]
Comment (nephila-nacrea) [reply templates/web/base/waste/bulky/booking_cancellation.html]: 👍 
Only thing I'm not sure about is 'It usually takes between x-x working days.'

File: perllib/FixMyStreet/Cobrand/Peterborough.pm [hunk 4]
```diff
@@ -733,6 +737,100 @@ sub bulky_per_item_costs {
...
+    )->parse_datetime($collection_date_str);
+
+    return $self->_check_within_bulky_cancel_window( $now_dt,
+        $collection_dt );
+}
+
+sub _check_within_bulky_cancel_window {
+    my ( undef, $now_dt, $collection_dt ) = @_;
+
+    # Collection date must be at least a day later than today's date for
```
Comment (dracos) [inline perllib/FixMyStreet/Cobrand/Peterborough.pm:805]: Unless I've missed something, I think this function can be done as:
```
my $collection_day = $collection_dt->clone->truncate( to => 'day' );
my $cutoff_dt = $collection_dt->clone->subtract( minutes => 5 );
return $now_dt < $cutoff_dt;
```

File: perllib/FixMyStreet/Cobrand/Peterborough.pm [hunk 5]
```diff
@@ -1267,6 +1365,27 @@ sub open311_contact_meta_override {
...
+    my ( $self, $data ) = @_;
+
+    my $c = $self->{c};
+
+    $data->{title}    = 'Bulky goods cancellation';
+    $data->{category} = 'Bulky cancel';
+    $data->{detail} .= " | Original report ID: "
+        . $c->stash->{property}{pending_bulky_collection}->id;
+
+    $data->{extra_COMMENTS} = 'Cancellation at user request';
```
Comment (dracos) [inline perllib/FixMyStreet/Cobrand/Peterborough.pm:1378]: Don't think we need to set this in $data, can just be set in the param next line.

**Review by dracos — CHANGES_REQUESTED** [review]
Looks good :) A few small changes but nothing major, ask if you have any questions, assuming the test being written tests the main flow I think is all that's missing.

File: perllib/FixMyStreet/Cobrand/Peterborough.pm (hunk 4 above)
Comment (nephila-nacrea) [reply perllib/FixMyStreet/Cobrand/Peterborough.pm:805]: Oh duh 🤦

File: perllib/FixMyStreet/Roles/SCP.pm [hunk 6]
```diff
@@ -117,6 +117,14 @@ sub garden_cc_check_payment_status {
 
     if ($resp->{transactionState} eq 'COMPLETE') {
         if ($resp->{paymentResult}->{status} eq 'SUCCESS') {
+            $p->set_extra_metadata( 'authCode',
+                $resp->{paymentResult}{paymentDetails}{authDetails}{authCode}
+            );
+            $p->set_extra_metadata( 'continuousAuditNumber',
+                $resp->{paymentResult}{paymentDetails}{authDetails}
+                    {continuousAuditNumber} );
+            $p->update;
```
[the remark this answers is not in the document]
Comment (nephila-nacrea) [reply perllib/FixMyStreet/Roles/SCP.pm:126]: Yeah, will keep it for now, but what would be a better solution?

File: perllib/FixMyStreet/App/Controller/Waste.pm [hunk 7]
```diff
@@ -1176,30 +1171,22 @@ sub process_bulky_data : Private {
...
 
     # Mark original report as closed
-    # XXX Leave e.g. comment to clarify that report is cancelled
-    $c->stash->{property}{open_bulky_collection_report}->state('closed');
-    $c->stash->{property}{open_bulky_collection_report}->update;
-
-    $c->forward( 'add_report', [ \%data ] ) or return;
+    $collection_report->state('closed');
+    $collection_report->detail(
+        $collection_report->detail . " | Cancelled at user request", );
```
[the remark this answers is not in the document]
Comment (nephila-nacrea) [reply perllib/FixMyStreet/App/Controller/Waste.pm:1188]: 👍

File: perllib/FixMyStreet/Cobrand/Peterborough.pm [hunk 8]
```diff
@@ -802,26 +803,20 @@ sub within_bulky_cancel_window {
 sub _check_within_bulky_cancel_window {
     my ( undef, $now_dt, $collection_dt ) = @_;
 
-    # Collection date must be at least a day later than today's date for
-    # cancellation to be allowed
-    my $today          = $now_dt->clone->truncate( to => 'day' );
     my $collection_day = $collection_dt->clone->truncate( to => 'day' );
```
Comment (dracos) [inline perllib/FixMyStreet/Cobrand/Peterborough.pm:806]: This value isn't used any more, so can be removed?

**Review by dracos — CHANGES_REQUESTED** [review]
Only two things - one line I don't think is needed any more, and there's no test of the free bulky reset code (as you can see on https://github.com/mysociety/fixmystreet/pull/4182/files looking at the Codecov warnings in perllib/Integrations/Bartec.pm and the calling code). A small extra test that cancels a FREE one and just checks that code is being called okay would be great, then think it's done :)

**Review by dracos — APPROVED** [review]
Looks good :) I've pushed a commit db67a94316f13c1cb43a13bc079d34a82736766d which hopefully is easy to understand - your new Bartec.pm function wasn't being tested at all as it was entirely mocked, I thought it would be better to mock the actual calls to Bartec but not the surrounding code which can then be tested by your existing test. Also it tidies up the response to make sure we always have an array without needing so much boilerplate checking. Let me know if you have any questions; assuming I haven't broken anything with that, ready to tidy up and merge in :)

**Comment by nephila-nacrea** [comment]
> Looks good :) I've pushed a commit [db67a94](https://github.com/mysociety/fixmystreet/commit/db67a94316f13c1cb43a13bc079d34a82736766d) which hopefully is easy to understand - your new Bartec.pm function wasn't being tested at all as it was entirely mocked, I thought it would be better to mock the actual calls to Bartec but not the surrounding code which can then be tested by your existing test. Also it tidies up the response to make sure we always have an array without needing so much boilerplate checking. Let me know if you have any questions; assuming I haven't broken anything with that, ready to tidy up and merge in :)

Thanks for this @dracos ; I'd thought I was mocking the actual calls to Bartec rather than the Bartec.pm function 🤦

**Pull request merged**